### ISSUE TITLE
Rootana Segfaulting Straight Away

### ISSUE THREAD

**Issue by AndrewEdmonds11** [alcap-org/AlcapDAQ#72]
The develop branch of rootana crashes straight away on Merlin (and on UCL desktop): 
The backtrace is:

```
===========================================================
There was a crash.
This is the entire stack trace of all threads:
===========================================================
#0  0x0000003c304ac8be in waitpid () from /lib64/libc.so.6
#1  0x0000003c3043e909 in do_system () from /lib64/libc.so.6
#2  0x00002b4f29b62998 in TUnixSystem::StackTrace (this=0x224ca50) at /gpfs/home/edmonds_a/AlcapDAQ/root/core/unix/src/TUnixSystem.cxx:2403
#3  0x00002b4f29b61813 in TUnixSystem::DispatchSignals (this=0x224ca50, sig=kSigSegmentationViolation) at /gpfs/home/edmonds_a/AlcapDAQ/root/core/unix/src/TUnixSystem.cxx:1279
#4  <signal handler called>
#5  GetTree (inFile=0x0, t_name=0x450066 "SetupTree") at src/main.cpp:301
#6  0x0000000000416354 in TSetupData::Instance () at src/main.cpp:319
#7  0x0000000000417452 in PrepareSingletonObjects () at src/main.cpp:379
#8  0x0000000000417c9e in main (argc=<value optimized out>, argv=<value optimized out>) at src/main.cpp:104
===========================================================
```

and it's because gInFile isn't being set in main.cpp (after grepping, it is only set to NULL at declaration and is passed as an argument to a few functions).

I don't know exactly want to start messing about in main.cpp in case it clashes with framework development so I will leave it to someone else (i.e. @benkrikler or @litchfld)

**Comment by benkrikler** [comment]
Sorry Andy, this is another one where i'm unable to reproduce this.  I'm on develop, just did `make clean; make` and things worked fine.  Grepping for 'gInFile' in main.cpp, I see the following:

``` cpp
34: static TFile *gInFile=NULL;
65:  gInFile = new TFile(arguments.infile);
66:  if(!gInFile->IsOpen()) {
68:    delete gInFile;
84:  TTree* eventTree = GetTree(gInFile,"EventTree");
121:  gInFile->Close();
282:  TTree* setupTree= GetTree(gInFile, "SetupTree");
```

So gInFile is definitely being set.

**Comment by benkrikler** [comment]
No wait, scrap that, I'd not pulled since this morning and now I'm seeing the problem.  

@litchfld, I see you've merged in work from feature/data_manager.  Are we expecting that rootana should work or is this a known issue?  Also, I'm a bit nervous that the network diagram doesn't show any evidence of you rebasing your branch from the latest version of develop.  Was this intentional?

**Comment by benkrikler** [comment]
So I've looked at the latest commits and it seems there's a fair bit of left over debugging stuff from Phill's work.  Other parts of main have been removed which is understandable if we're trying to debug this Event Navigator, but this means the rest of rootana is unuseable. 

 @litchfld,  I think we need to make sure others can develop in parallel for now, so I suggest we roll back changes to main.cpp and hook up the event navigator on a parallel feature branch.  Then we merge it back in once we're confident it's working.

**Comment by litchfld** [comment]
Whups sorry, must have made more changes than I remembered. Should have just cherry-picked the old main before finishing. 

Sure, go ahead and roll back main.cpp.  It doesn't actually need to change it 'till everything else is done.

**Comment by litchfld** [comment]
Try now. I think I reverted successfully, although I couldn't work out the git way of doing it so I did it the cvs way.

**Comment by benkrikler** [comment]
Holy crap!  We've got ourselves one hell of a network diagram!  I think the git way of doing this would have been to `git checkout commit_where_main_was_good -- src/main.cpp`. And I would have just done this on an up-to-date copy of develop and pushed it directly, so it looks like a simple single commit.  I think we've said that for bug fixes on develop that can be done in one simple commit just do them on develop directly.

I've just checked and it is now working.  If @AndrewEdmonds11 is ok, I think we can close this issue then.

**Comment by litchfld** [comment]
>  I think the git way of doing this would have been to `git checkout commit_where_main_was_good -- src/main.cpp`

Ahah... I clearly don't think in the same way as the git developers. I was looking at cherry-pick.  I thought 'checkout' just meant switch branches in git-speak.  Ugh.

**Comment by litchfld** [comment]
We can stop integrating blame on me now; this is fixed.

**Comment by jrquirk** [comment]
I was trying to export a pulse using the following code (I made a module called Test)

```
int Test::ProcessEntry(TGlobalData* gData,TSetupData *setup){
  if (gData->fPulseIslandToChannelMap.count("Nc82") == 0)
    return 0;
  std::vector<TPulseIsland*>& tpis = gData->fPulseIslandToChannelMap.at("Nc82");
  for (unsigned int i = 0; i < tpis.size(); ++i)
    if (tpis[i]->GetSamples()[tpis[i]->GetPeakSample()] > 2700)
      if (ExportPulse::Instance())
    ExportPulse::Instance()->AddToExportList(TSetupData::Instance()->GetDetectorName("Nc82"), i);
  return 0;
}
```

I'm just looking at run 3200, hence the hardcoded values. My modules file was

```
debug=all
[ MODULES ]
draw_pulses = ExportPulse
test = Test
```

And I ran

```
$ ./rootana -i ~/data/tree/tree03200.root -o out03200.root -m src/testModule.txt -n 100
```

and then there was a crash

```
===========================================================
There was a crash.
This is the entire stack trace of all threads:
===========================================================
#0  0x00007f98c71cc46a in __libc_waitpid (pid=6597, stat_loc=stat_loc
entry=0x7fff1b34e310, options=options
entry=0) at ../sysdeps/unix/sysv/linux/waitpid.c:31
#1  0x00007f98c7150fae in do_system (line=0x163ce50 "/home/jrquirk/.local/root_5.34.07/etc/gdb-backtrace.sh 6594 1>&2") at ../sysdeps/posix/system.c:148
#2  0x00007f98cd13ee6d in TUnixSystem::StackTrace() () from /home/jrquirk/.local/root_5.34.07/lib/libCore.so.5.34
#3  0x00007f98cd141923 in TUnixSystem::DispatchSignals(ESignals) () from /home/jrquirk/.local/root_5.34.07/lib/libCore.so.5.34
#4  <signal handler called>
#5  0x00007f98c74cc7c8 in main_arena () from /lib/x86_64-linux-gnu/libc.so.6
#6  0x0000000000418920 in ClearGlobalData (data=0x1445610) at src/main.cpp:230
#7  0x0000000000418e0d in Main_event_loop (dataTree=dataTree
entry=0x1437240, arguments=...) at src/main.cpp:178
#8  0x0000000000415934 in main (argc=<optimized out>, argv=<optimized out>) at src/main.cpp:114
===========================================================


The lines below might hint at the cause of the crash.
If they do not help you then please submit a bug report at
http://root.cern.ch/bugs. Please post the ENTIRE stack trace
from above as an attachment in addition to anything else
that might help us fixing this issue.
===========================================================
#5  0x00007f98c74cc7c8 in main_arena () from /lib/x86_64-linux-gnu/libc.so.6
#6  0x0000000000418920 in ClearGlobalData (data=0x1445610) at src/main.cpp:230
#7  0x0000000000418e0d in Main_event_loop (dataTree=dataTree
entry=0x1437240, arguments=...) at src/main.cpp:178
#8  0x0000000000415934 in main (argc=<optimized out>, argv=<optimized out>) at src/main.cpp:114
===========================================================
```

The lines above might hint at the cause of the crash. I checked out develop immediately before, and did `make`s and `make clean`s, and I did hard resets of my head. But, as I've lost faith that this is enough to get rid of any problems I've introduced, I went to my home computer and did a fresh clone of the repo (but used my own version of ROOT) and tried running on develop (with the testModule.txt file in develop) and got the exact same error.

@litchfld @benkrikler please help. I'll try to look into it, but there's quite a bit of history to look through I'm not familiar with.

**Comment by jrquirk** [comment]
TGlobalData::Clear() should replace the line

```
std::vector<TPulseIsland*> pulse_vector= mapIter->second;
```

with

```
std::vector<TPulseIsland*>& pulse_vector= mapIter->second;
```

for the line

```
pulse_vector.clear()
```

To have any effect. I haven't pushed this change because there are other errors now (bad_alloc if I run more than a few tens of blocks). Additionally, can we get rid of the ClearGlobalData(TGlobalData*) global function in main.cpp? I think the answer is yes, but I'm not certain so I won't delete it.

**Comment by benkrikler** [comment]
This should probably be in a separate issue to the original because it's not the same underlying problem.

> GlobalData::Clear() should replace the line
> 
> ```
> std::vector<TPulseIsland*> pulse_vector= mapIter->second;
> ```
> 
> with
> 
> ```
> std::vector<TPulseIsland*>& pulse_vector= mapIter->second;
> ```

That's a true statement in general probably; the top line copies the original vector (ie. internally may be looping over each element), the second one just makes a reference to it, so it shares the same memory.  ( I know you know this, John, but I've seen this sort of thing so much in our code that I want to point it out).

> for the line
> 
> ```
> pulse_vector.clear()
> ```
> 
> To have any effect.

That's definitely true and probably the intention in this function. 

I'm not sure that this is the cause of this problem though. Your backtrace mentions line 230 as the cause:

> ```
> #6  0x0000000000418920 in ClearGlobalData (data=0x1445610) at src/main.cpp:230
> ```

which is the line with `delete pulse_vector[i];` in it.  I checked that it wasn't because pulses were being deleted but left in the vector as `pulseIter.clear()` doesn't affect the original one, like you say.  I checked the size and capacity of the vectors before the pulses are deleted with and without the `&` and they don't change.  So it seems root is doing something more intelligent each time it fills these vectors with the next event than just putting the new set of pulses into the container but otherwise leaving it unchanged.

Given this, I think the cause of this might actually be somewhere else in the code deleting TPIs perhaps? It would be nice to run this through valgrind but with so much ROOT code in there, I'm not sure if it would be very instructive.

> Additionally, can we get rid of the ClearGlobalData(TGlobalData*) global function in main.cpp?

I think the event navigator will absorb this responsibility pretty soon, so lets leave this in until then.

**Comment by benkrikler** [comment]
Just to check things,  I created a module called Test and filled it's ProcessEntry method with your code.  I only have run 3329 locally, so I changed the limit from 2700 to 200.  

I ran with 100 events, and everything worked fine. I'll try with your run and see if I get the same problem.  Maybe that tree somehow contains bad events.

**Comment by benkrikler** [comment]
Having done `make clean; make` in both my alcapana work directory and rootana, I now see this bug.  I've jumped back through the commit history and so far have found that it broke somewhere between 1bc4ad9a36aab02eff5430ce601d68bd229b99a5 where it worked, and e1ef69c9a6141eb4808fdeab6b1c81702d775732 where it was broken.  

Time to try out git bisect...

**Comment by benkrikler** [comment]
So git bisect is completely awesome!  I followed this tutorial: https://www.kernel.org/pub/software/scm/git/docs/git-bisect.html and wrote a small script to run the test:

``` bash
#/bin/bash 

set -e

function Rebuild(){
cd "$1"
echo
echo ==== Now in : `pwd`
make clean
make -j4
}

Rebuild $DAQdir/analyzer/work/krikler_b
Rebuild $DAQdir/analyzer/rootana

./rootana -i ~quirk_j/data/tree/tree03200.root -o output.root -m src/tmp_module -n 10
```

where the modules file tmp_module was:

``` cfg
debug=all
[ MODULES ]
draw_pulses = ExportPulse

[draw_pulses]
SiR2-S = 1(0)
```

I was already at the good commit (1bc4ad9a36aab02eff5430ce601d68bd229b99a5) so to get going I issued:

``` git
git bisect start
git good 
git bad e1ef69c9a6141eb4808fdeab6b1c81702d775732
```

It took about 15 minutes and 7 attempts to find the culprit (283c9047546cb1025059ecc87eb266752c311a96):

```
283c9047546cb1025059ecc87eb266752c311a96 is the first bad commit
```

although the commit before it (0c3944d8153847c8e91bd81aa7e5b37b35ff5040) failed to compile so I marked it as good and moved on.

This was, it turns out, the last time where rootana seg-faulted immediately (yup, turns out we're on the right ticket! ).  So at that point, something else must have changed that went unnoticed since.

I'll look into what that could have been, but for longevity, here's the output of `git bisect log` which can be used with `git bisect replay` to redo this more quickly:

``` bash
git bisect start
# good: [1bc4ad9a36aab02eff5430ce601d68bd229b99a5] Fixed bug in Template_module file
git bisect good 1bc4ad9a36aab02eff5430ce601d68bd229b99a5
# bad: [e1ef69c9a6141eb4808fdeab6b1c81702d775732] Change TriggerPlotter to PulseViewer in testModule file
git bisect bad e1ef69c9a6141eb4808fdeab6b1c81702d775732
# good: [8a1a7af7170c0d432a5d45749af76d6b1fba86b6] Everything is implemented in TriggerPlotter now, can begin run debugging
git bisect good 8a1a7af7170c0d432a5d45749af76d6b1fba86b6
# good: [7f6603e7f140685738aafd60312a9dd9532f0853] Ths should complete most of the interface for the event navigator (to zeroth order) Aim is to merge the branch back into develop and then start a new feature where we (a) make it conform to Ben's new IDs, then (b) write the guts.
git bisect good 7f6603e7f140685738aafd60312a9dd9532f0853
# good: [57a4cb1a78d0df7821f90f280e00366b0705f87a] Correctly handle replacing options value
git bisect good 57a4cb1a78d0df7821f90f280e00366b0705f87a
# good: [5c16058d5163b08a9394b4c25fb901cf614df1b9] AE: Removing one of my debug messages
git bisect good 5c16058d5163b08a9394b4c25fb901cf614df1b9
# good: [e8e4ce2654cb9495a1029e44d37c8cb45f79ce88] AE: Rebasing before finishing. Merge branch 'feature/AE_LLDQ_plot_changes' of github.com:alcap-org/AlcapDAQ into feature/AE_LLDQ_plot_changes
git bisect good e8e4ce2654cb9495a1029e44d37c8cb45f79ce88
# good: [7e345b3dd50620794636ca69937ac61ace07b37e] AE: I went out of date quickly. Merge branch 'develop' of github.com:alcap-org/AlcapDAQ into develop
git bisect good 7e345b3dd50620794636ca69937ac61ace07b37e
# bad: [0c3944d8153847c8e91bd81aa7e5b37b35ff5040] Merge branch 'develop' of github.com:alcap-org/AlcapDAQ into develop
git bisect bad 0c3944d8153847c8e91bd81aa7e5b37b35ff5040
# bad: [283c9047546cb1025059ecc87eb266752c311a96] Revert main.cpp to last version on develop before merge of features: Seems I merged in some debug code without realising.
git bisect bad 283c9047546cb1025059ecc87eb266752c311a96
```

**Comment by jrquirk** [comment]
I'm on a phone so I can't be the most descriptive, but there was a double delete problem because TPI::Clear was being called right before ::ClearGlobalData, and the TPIs were getting deleted but the vector of pointers was not due to the lack of the ampersand pointed out. Hence the problem at main.cpp:230 (the second delete).

**Comment by benkrikler** [comment]
I'm not sure what you mean by `TPI::Clear` cause I can't see any function like that.  As far as I can see, TPIs are only 'deleted' within ClearGlobalData.

**Comment by litchfld** [comment]
Short answer... err I dunno.

Long answer: I think this may be related to an investigation I was doing over Easter.   Pull up a seat.  

### In our code

There is an 'issue' running on systems with a low virtual memory limit (~1GiB). For some reason rootana has a tendency to run up to 2 or 3GiB after a few tens of events. It shrinks back down again afterwards, but that is too late if the kernel refused the process  memory - there will be a std::bad_alloc thrown, which may be the cause of what John saw on his home machine.

I discovered this over Easter because the normal work machines were powered off, so i switched to the gateway, and spent a week trying to work out what I had done before discovering there was a lower limit on that machnie.

So I tried a lot of fiddling with memory Clears and such.  I don't think any of those changes persisted into the code base - there's no evidence in the partitioned history.  but I need to check more widely to be sure.

I'd quite like to get rid of the ClearGlobalData() function in any case - it's a dubious idiom. The sane thing is for TGlobalData to clear itself, since it needs to know how to do that in the destructor anyway.  I originally implemented it this way, but Michael pointed out we diddn't know what affect it would have in the online.  Thinking about it the answer should be none, because no-one uses it online. 

### What root might be doing

>  So it seems root is doing something more intelligent each time it fills these vectors with the next event than just putting the new set of pulses into the container but otherwise leaving it unchanged.

Probably this is related to the [branch ownership rules](http://root.cern.ch/root/html/TBranchElement.html#TBranchElement:SetAddress) which are complicated.  I'll take a look.

### Cleanup and copies

> >    GlobalData::Clear() should replace the line
> > `std::vector<TPulseIsland*> pulse_vector= mapIter->second;` with
> > `std::vector<TPulseIsland*>& pulse_vector= mapIter->second;`
> 
> That's a true statement in general probably; the top line copies the original vector (ie. internally may be looping over each element), the second one just makes a reference to it, so it shares the same memory. ( I know you know this, John, but I've seen this sort of thing so much in our code that I want to point it out).

_Digression: A habit that I increasingly like (but don't always follow) is to make the copy constructor and copy assignment operator private for every class until such time as you have discovered a case when you_ need _to make copies. At that point you should hopefully know what copy semantics are suitable._

Anyway, it is true in general although I'm not sure whether it matters here.  The pointers are all deleted which releases the majority of the memory.  ROOT then overwrites the STL class somehow which may or ~~may not~~ involve clearing, overwriting or deleting the existing vector. But i'm pretty sure (see [another old roottalk post](http://root.cern.ch/root/roottalk/roottalk00/2313.html) it does some kind of clean-up.  I've never wanted to go searching for the STL streamer classes, and I don't remember seeing anything explicit about what it does do.

**Comment by litchfld** [comment]
One thing that might be useful. I added a file [debug_tools.h](https://github.com/alcap-org/AlcapDAQ/blob/develop/analyzer/rootana/src/debug_tools.h) which provides a few quick and dirty routines to look at a jobs memory and time profile. e.g. 

``` C++
std::cout << DEBUG::check_mem().str << endl;
```

will print the current memory usage.  You can instead make a graph of the time evolution, whatever you like.

**Comment by jrquirk** [comment]
@benkrikler 

> I'm not sure what you mean by `TPI::Clear`

I meant TGlobalData::Clear, my phone autocorrected to TPI.

@litchfld 

> Anyway, it is true in general although I'm not sure whether it matters here.

I guess I meant in the case of avoiding trying to `delete NULL`, I have no idea how ROOT handles loading/unloading from memory. For a check can we delete gData as well, thereby calling the vector destructor, before having ROOT grab the next event?

**Comment by litchfld** [comment]
`delete NULL`  is perfectly allowable though (the program knows not to do anything).  So I think we can delete the whole of TGlobalData ourselves if we like, it should correspond to option 3 discussed in SetBranchAddress

**Comment by benkrikler** [comment]
We need a proper fix for this.  @litchfld, you're saying we should delete TGlobalData at the end of each iteration of the event loop?

**Comment by litchfld** [comment]
I'm not sure...  failing to delete stuff should not cause a SEGV, (it will eventually be a `std::bad_alloc` though)

Sorry i haven't had a chance to look at this yet, Saturday I got some weird depends-on-your-libstdc++ error, and today and tomorrow i'm at a T2K meeting so only working ~10 mins in an hour.  If you think you have a fix...

**Comment by litchfld** [comment]
OK I got the HEAD of develop running with the default `testModule.txt` file.  But the error I see is completely different.

```
0x00002aaaaaf845bc in TBuildRealData::Inspect (this=0x7fffffffdd60, 
    cl=0xe78280, pname=0xe77e70 "", mname=0x2aaaafaab844 "first", add=0xe61ea0)
    at /unix/muons/users/litchfld/alcap/AlcapDAQ/root/core/meta/src/TClass.cxx:529
529           if (!fRealDataClass->InheritsFrom(cl)) {
```

I think we've seen this before, which I'm going to look into. But given the backtrace it may just be i'm using an old root file or some incompatible combination of code/modules/datafile.  

Can someone who can reproduce the error put a root file [could be a truncated tree if too large] and a Module file in dropbox  (and anything else i've forgotten) so I can see if that helps me reproduce the error?

**Comment by AndrewEdmonds11** [comment]
I recently copied tree02808.root from our second production run to UCL. The path is /unix/muons/alcap/data/tree. It's the one I've been using and crashes on the develop branch, using testModule.txt, with the following error:

```
 *** Break *** segmentation violation



===========================================================
There was a crash.
This is the entire stack trace of all threads:
===========================================================
#0  0x00000031f169a075 in waitpid () from /lib64/libc.so.6
#1  0x00000031f163c741 in do_system () from /lib64/libc.so.6
#2  0x00002b58211ad782 in TUnixSystem::StackTrace (this=0x17d47270)
    at /unix/muons/users/edmonds/AlCap/AlcapDAQ/root/core/unix/src/TUnixSystem.cxx:2403
#3  0x00002b58211ab03a in TUnixSystem::DispatchSignals (this=0x17d47270, 
    sig=<value optimized out>)
    at /unix/muons/users/edmonds/AlCap/AlcapDAQ/root/core/unix/src/TUnixSystem.cxx:1279
#4  <signal handler called>
#5  0x000000001854ef60 in ?? ()
#6  0x000000000041a180 in ClearGlobalData (data=0x184fa5b0) at src/main.cpp:241
#7  0x000000000041a66d in Main_event_loop (dataTree=0x184ec5d0, arguments=Unhandled dwarf expression opcode 0xf3
    at src/main.cpp:189
#8  0x000000000041b94f in main (argc=Unhandled dwarf expression opcode 0xf3
===========================================================


The lines below might hint at the cause of the crash.
If they do not help you then please submit a bug report at
http://root.cern.ch/bugs. Please post the ENTIRE stack trace
from above as an attachment in addition to anything else
that might help us fixing this issue.
===========================================================
#5  0x000000001854ef60 in ?? ()
#6  0x000000000041a180 in ClearGlobalData (data=0x184fa5b0) at src/main.cpp:241
#7  0x000000000041a66d in Main_event_loop (dataTree=0x184ec5d0, arguments=Unhandled dwarf expression opcode 0xf3
    at src/main.cpp:189
#8  0x000000000041b94f in main (argc=Unhandled dwarf expression opcode 0xf3
===========================================================
```

Hope that helps

**Comment by litchfld** [comment]
Sad times... that's exactly the combination I tried. Better check my build history.

**Comment by litchfld** [comment]
OK. I think I fixed it. As John spotted,it does seem to be a simple case of double deletes in TGlobalData::Clear() and ClearGlobalData(), caused by the accidental copy.  I haven't check out why Ben's git bisect points to that `Clear("C")` change although I have a suspicion. 

It might be nice if we could disallow copying of `std::vector<TXxxPulses*>`, but of course that is not possible.  One might consider a simple change:

``` C++
class TXxxPulseList: public std::vector<TXxxPulseList*>
{
private:
TXxxPulseList::TXxxPulseList(const TXxxPulseList& src){}
//similarly for copy asm't operator
};
```

But this doesn't really help if people are still in the habit of using `std::vector<TXxxPulseList*>` directly.

**Comment by benkrikler** [comment]
Since we already do essentially this by providing typedefs in the definitions file that suggestion wouldn't be so hard. And since the event navigator itself would store this derived class, it wouldn't be possible to assign or copy construct a `std::vector<TXxxPulseList*>` from it.
If we do sub-class `vector<pulse*>` I could overload push_back to tell the pulse its position in the list (ie. push_back sets the pulse's ID ). Not that I want to revisit that topic.....

**Comment by thnam** [comment]
I don't know if this is related, but I could not run rootana either on develop or release/data_quality_3.1. The error messages are:

```
Problem opening modules file: H���5�%
Problem setting up analysis modules.
```

**Comment by AndrewEdmonds11** [comment]
Hi Nam. That's the same as issue #102. I don't think they're related but they should both be fixed soon.

**Comment by thnam** [comment]
I see, thank you.

**Comment by benkrikler** [comment]
I think that problem is the one Phil has described and fixed or is fixing on ticket #102..
Sent from my mobile

**Comment by litchfld** [comment]
@thnam I pushed a fix to branch feature/RPL_command_line, which will merge into develop soon.  See #102.

**Comment by thnam** [comment]
@litchfld Thanks a lot.

**Comment by ajidaniel** [comment]
I am having problems while running rootana (was trying to go through Ben's walkthrough). Checked out software from git from the master branch today, compiled every thing  and ran rootana on merlin for run 3200.  

While I was going through the issues it sounds more or less similar to what is discussed here, I think.  I am not that familiar cpp to fix it myself. Was this resolved ?. Could someone point me the relevant entries (I tried a fresh checkout to my local machine and I am getting the same error message)
Following is the relevant part of the  output.

There are 4523 entries
Processing file, which may take a while. Have patience young padawan..
Will analyse 2 of 28 channels
SiR2-S: using generator: MaxBin
SiR2-F: using generator: MaxBin

 **\* Break **\* segmentation violation

There was a crash.
This is the entire stack trace of all threads:

#0  0x00000037f62ac8be in waitpid () from /lib64/libc.so.6
#1  0x00000037f623e909 in do_system () from /lib64/libc.so.6
#2  0x00002ae63eb31898 in TUnixSystem::StackTrace (this=0x195e060) at /gpfs/home/daniel_a/AlcapDAQ/root/core/unix/src/TUnixSystem.cxx:2403
#3  0x00002ae63eb30713 in TUnixSystem::DispatchSignals (this=0x195e060, sig=kSigSegmentationViolation) at /gpfs/home/daniel_a/AlcapDAQ/root/core/unix/src/TUnixSystem.cxx:1279
#4  <signal handler called>
#5  0x00000000021e9cb0 in ?? ()
#6  0x0000000000417a73 in ClearGlobalData (data=0x218fcc0) at src/main.cpp:241
#7  0x0000000000418ac5 in Main_event_loop (dataTree=0x2181880, arguments=<value optimized out>) at src/main.cpp:189
#8  0x000000000041923c in main (argc=<value optimized out>, argv=<value optimized out>) at src/main.cpp:125*

**Comment by benkrikler** [comment]
Hey Aji, so you're running on the master branch?  Try to checkout the develop branch (`git checkout develop`) and then rebuild everything (`make clean; make` for both alcapana and rootana).  Does it still crash there?

**Comment by benkrikler** [comment]
I've been able to confirm this crash on the `master` branch, but running on `develop` seems ok to me.  I'm guessing we need to merge into master the changes on develop that fixed this problem.  On the whole we seem to be running with develop and saving master for production I think.

**Comment by ajidaniel** [comment]
Hi Ben,
I did checkout develop, installed sqlite by excecuting install-sqlite3.sh, properly gave the path as defined in thisdaq.sh.  However, am getting a complaint about TSQLiteServer.h while compiling rootana.

 src/TAP_generators/utils/PulseCandidateFinder.cpp:10:27: fatal error: TSQLiteServer.h: No such file or directory
 #include <TSQLiteServer.h>
                          ^
compilation terminated.

Did I miss some intermediate step?
Thanks
Aji.

**Comment by benkrikler** [comment]
I changed the root install script as well last night.  Installing root from scratch takes a while though, but I think if you change into `root-build` and do:

```
cmake -Dsqlite=on -DSQLITE_INCLUDE_DIR=../sqlite-install/include -DSQLITE_LIBRARIES=../sqlite-install/lib/libsqlite3.so  .
```

you could then rebuild th sqlite libraries only by doing `make; make install` in the same directory (`root-build`).

**Comment by ajidaniel** [comment]
Ben I did compile root as you mentioned in your previous entry. Still rootana has some struggle finding some relevant headers. This time a different message though.  If you have time we can discuss this in today's meeting.

Thanks.
Following is the relevant o/p from rootana compilation

```
Making object: 'obj/PulseCandidateFinder.o'
In file included from src/TAP_generators/utils/PulseCandidateFinder.cpp:10:0:
/home/adaniel/AlCap/AlcapDAQ/root-install/include/TSQLiteServer.h:20:21: fatal error: sqlite3.h: No such file or directory
 #include <sqlite3.h>
                     ^
compilation terminated.
make: *** [obj/PulseCandidateFinder.o] Error 1
```

**Comment by benkrikler** [comment]
Can you try pulling Aji and recompiling?  Make sure to source the thisdaq.sh script first, but hopefully we should now be specifying how to find the sqlite files properly...

**Comment by ajidaniel** [comment]
Hi Ben, after today's meeting I pulled it from develop branch, recompiled everything ( including a fresh  recompilation of root ). However, I am getting the same error as in my previous post when I attempt to compile rootana.

**Comment by benkrikler** [comment]
I just (~20 minutes ago) pushed a slight change to the Makefile though, have you pulled develop since then?  Also, did you run the install-sqlite.sh script?

**Comment by ajidaniel** [comment]
Nope, this was about 2hrs ago. Also, I didn't do a fresh installation of sqlite.  I will pull it now, do a new installation of sqlite, and wil let you know.
Thanks

**Comment by ajidaniel** [comment]
Ben, after doing a fresh checkout from develop branch, and a fresh installation of everything(root, sqlite, ..) solved the problem with rootana compilation. I have also managed to do a test run of rootana with modified testModule.txt
Thanks for the help.
Aji.

**Comment by benkrikler** [comment]
That's great news!! 

I'm going to close this ticket then.  When things next break we can open it again...
